File: Daily-Streak/Day-06_Maximum-Level-Sum-of-a-Binary-Tree/Solution.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public int maxLevelSum(TreeNode root) {
        // ✅ OPTIMAL APPROACH: BFS

        Queue<TreeNode> queue = new LinkedList<>();
        queue.offer(root);

        int level = 1;
        int answerLevel = 1;
        long maxSum = Long.MIN_VALUE;

        while (!queue.isEmpty()) {
            int size = queue.size();
            long currentSum = 0;

            // Process current level
            for (int i = 0; i < size; i++) {
                TreeNode node = queue.poll();
                currentSum += node.val;

                if (node.left != null)  queue.offer(node.left);
                if (node.right != null) queue.offer(node.right);
            }

            if (currentSum > maxSum) {
                maxSum = currentSum;
                answerLevel = level;
            }

            level++;
        }

        return answerLevel;
    }

    /*
    ===========================
    ❌ DFS APPROACH (COMMENTED)
    ===========================
    - Track depth manually
    - Use HashMap<Integer, Integer> for sums
    - More complex than BFS

    ===========================
    🧪 STANDALONE VERSION (COMMENTED)
    ===========================
    public static void main(String[] args) {
        TreeNode root = new TreeNode(1);
        root.left = new TreeNode(7);
        root.right = new TreeNode(0);
        root.left.left = new TreeNode(7);
        root.left.right = new TreeNode(-8);

        Solution s = new Solution();
        System.out.println(s.maxLevelSum(root));
    }
    */
}

File: Daily-Streak/Day-06_Maximum-Level-Sum-of-a-Binary-Tree/Solution.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int maxLevelSum(TreeNode* root) {
        // ✅ OPTIMAL APPROACH: BFS (Level Order Traversal)

        queue<TreeNode*> q;
        q.push(root);

        int level = 1;
        int answerLevel = 1;
        long long maxSum = LLONG_MIN;

        while (!q.empty()) {
            int size = q.size();
            long long currentSum = 0;

            // Process all nodes of current level
            for (int i = 0; i < size; i++) {
                TreeNode* node = q.front();
                q.pop();

                currentSum += node->val;

                if (node->left)  q.push(node->left);
                if (node->right) q.push(node->right);
            }

            // Update max sum and level
            if (currentSum > maxSum) {
                maxSum = currentSum;
                answerLevel = level;
            }

            level++;
        }

        return answerLevel;
    }

    /*
    ===========================
    ❌ DFS APPROACH (COMMENTED)
    ===========================
    - Use DFS and track depth
    - Store sums in map or vector
    - More bookkeeping required
    - BFS is cleaner for level-based problems

    ===========================
    🧪 STANDALONE VERSION (COMMENTED)
    ===========================
    #include <bits/stdc++.h>
    using namespace std;

    int main() {
        TreeNode* root = new TreeNode(1);
        root->left = new TreeNode(7);
        root->right = new TreeNode(0);
        root->left->left = new TreeNode(7);
        root->left->right = new TreeNode(-8);

        Solution s;
        cout << s.maxLevelSum(root) << endl;
        return 0;
    }
    */
};
